### PR TITLE
feat: support 192/128 and 256/256 ffa_fa4 in magi

### PR DESCRIPTION
**Pull request by NVIDIA-JerryChen**
## Summary

  - Add `head_dim_v` propagation and support SM100 configurations:
    - Q/K dim = 192, V dim = 128
    - Q/K/V dim = 256
  - Align forward q-stage selection and backward K2Q CSR metadata with the FFA_FA4 kernel tile contract.

  ## KV Communication

  - Symmetric `(256, 256)` path keeps the existing sequence-axis fusion with `packed_times=2`.
  - Asymmetric `(192, 128)` path uses MLA-style K/V packing:
    - concatenate K and V along the last head-dimension axis;
    - communicate them as one fused token stream with `packed_times=1`;
    - split the received buffer back into K and V.
  - Backward dK/dV communication follows the same fused representation and is split after reduction.
  - The asymmetric path does not use the symmetric `concat_kv`/`concat_dkv` sequence-axis packing.

  ## CSR Tile Mapping

  - `(192, 128)` backward K2Q CSR metadata uses `128x256` blocks.
  - `(256, 256)` backward K2Q CSR metadata uses `256x128` blocks.
  - Forward q-stage and CSR block sizes are selected to match the FFA_FA4 kernel.

  ## Validation

  - Ran selected SM100 FFA_FA4 tests for configurations with `n_func <= 13`.
  
  ## Related PR

  - Supersedes [MagiAttention#326](https://github.com/SandAI-org/MagiAttention/pull/326).
  - After this PR is merged, PR #326 can be closed.

### CODE REVIEW

**Review by lijinnn — COMMENTED** [review]
Hello! Excited to work together. 🙌

I noticed that the pre-commit checks failed for this PR: https://github.com/SandAI-org/MagiAttention/pull/331.

Could you please refer to our contributing guide (https://github.com/SandAI-org/MagiAttention/blob/main/CONTRIBUTING.md) to run the pre-commit checks locally, fix the formatting, and update your commit?

Thank you so much!

File: magi_attention/functional/dist_attn.py [hunk 1]
```diff
@@ -2089,8 +2146,34 @@ def _reduce_partial_dkv(
...
+                    partial_local_dkv, dim=-1
+                )  # cat-on-last-dim is contiguous
+            partial_local_dkv = self._asym_fused_local_dkv
+            if isinstance(partial_remote_dkv, tuple):
+                partial_remote_dkv = torch.cat(partial_remote_dkv, dim=-1)
+            ref_remote_dkv = torch.cat(ref_remote_dkv, dim=-1)
+
+        # The fused-buffer path covers both symmetric concat_dkv=True and the
+        # asymmetric cat-on-feature case.
+        is_fused = self.concat_kv or asymmetric_kv
```
Comment (lijinnn) [inline magi_attention/functional/dist_attn.py]: Could you change this to use self.concat_dkv instead? It would be a better approach.

File: magi_attention/functional/dist_attn.py
Comment (lijinnn) [inline magi_attention/functional/dist_attn.py]: In _no_overlap_backward, the dkv_shape computation assumes symmetric K/V:

dkv_shape = (full_k.shape[0] * 2, *full_k.shape[1:])
When head_dim_v != head_dim (e.g. the new 192/128 config), full_k.shape[-1] is d_k (192) but full_v.shape[-1] is d_v (128). The fused dkv buffer would be allocated entirely with d_k as the last dimension, giving dV the wrong shape.

This path isn't exercised by the current tests (which use overlap mode), but it would break if someone runs no_overlap=True with asymmetric K/V. Might be worth either fixing it or adding an explicit guard:

assert not dist_attn_runtime.is_asymmetric_kv, (
"no_overlap mode does not yet support asymmetric K/V (head_dim != head_dim_v)"
)

Comment (NVIDIA-JerryChen) [reply magi_attention/functional/dist_attn.py]: Thanks for pointing this out. I removed dkv_shape from the _launch_attn_bwd_kernel interface because the FFA_FA4 and SDPA paths return separate dK and dV tensors and do not use a fused dKV allocation. The legacy fused-buffer backend now derives dkv_shape internally and explicitly requires symmetric K/V shapes. This also removes the ambiguous symmetric-shape assumption from the no-overlap FFA_FA4 path.

File: magi_attention/functional/dist_attn.py (hunk 1 above)
Comment (NVIDIA-JerryChen) [reply magi_attention/functional/dist_attn.py]: Changed the backward fused-buffer condition to use self.concat_dkv

**Review by lijinnn — APPROVED** [review]
LGTM
